edb/pgsql/compiler/clauses.py

```diff
@@ -77,6 +77,58 @@ def fini_stmt(
                 )
 
 
+def get_volatility_ref(
+        path_id: irast.PathId, *,
+        ctx: context.CompilerContextLevel) -> pgast.BaseExpr:
+    """Produce an appropriate volatility_ref from a path_id."""
+
+    ref: Optional[pgast.BaseExpr] = relctx.maybe_get_path_var(
+        ctx.rel, path_id, aspect='identity', ctx=ctx)
+    if not ref:
+        rvar = relctx.maybe_get_path_rvar(
+            ctx.rel, path_id, aspect='value', ctx=ctx)
+        if rvar and isinstance(rvar.query, pgast.ReturningQuery):
+            # If we are selecting from a nontrivial subquery, manually
+            # add a volatility ref based on row_number. We do it
+            # manually because the row number isn't /really/ the
+            # identity of the set.
+            name = ctx.env.aliases.get('key')
+            rvar.query.target_list.append(
+                pgast.ResTarget(
+                    name=name,
+                    val=pgast.FuncCall(name=('row_number',), args=[],
+                                       over=pgast.WindowDef())
+                )
+            )
+            ref = pgast.ColumnRef(name=[rvar.alias.aliasname, name])
+        else:
+            ref = relctx.get_path_var(
+                ctx.rel, path_id, aspect='value', ctx=ctx)
+
+    return ref
+
+
+def setup_iterator_volatility(
+        iterator: Optional[Union[irast.Set, pgast.IteratorCTE]], *,
+        is_cte: bool=False,
+        ctx: context.CompilerContextLevel) -> None:
+    if iterator is None:
+        return
+
+    old = () if is_cte else ctx.volatility_ref
+
+    path_id = iterator.path_id
+    ref: Optional[pgast.BaseExpr] = None
+
+    def get_ref() -> pgast.BaseExpr:
```

Comment:
Can you explain why a callback approach is needed, please?

Reply:
I touched on it in the comment I added in `context.py` but could add a comment here too: it's so we don't add the volatility ref column to the relation unless it actually gets used.